spec/std/software_version_spec.cr

```diff
@@ -0,0 +1,168 @@
+    SoftwareVersion.valid?("an invalid version").should be_false
+  end
+
+  it "equals" do
+    assert_version_equal "1.2", "1.2"
+    assert_version_equal "1.2.b1", "1.2.b.1"
+    assert_version_equal "1.0+1234", "1.0+1234"
+    refute_version_equal "1.2", "1.2.0"
+    refute_version_equal "1.2", "1.3"
+  end
+
+  it ".new with number" do
+    assert_version_equal "1", SoftwareVersion.new(1)
+    assert_version_equal "1.0", SoftwareVersion.new(1.0)
+  end
+
+  it ".new with segments" do
+    SoftwareVersion.new(1).to_s.should eq "1"
+    SoftwareVersion.new(1, 0).to_s.should eq "1.0"
+    SoftwareVersion.new(1, 0, 0).to_s.should eq "1.0.0"
+    SoftwareVersion.new(1, 0, 0, prerelease: "rc1",).to_s.should eq "1.0.0-rc1"
+    SoftwareVersion.new(1, 0, 0, prerelease: "rc1", metadata: "build8").to_s.should eq "1.0.0-rc1+build8"
+    SoftwareVersion.new(1, prerelease: "rc1",).to_s.should eq "1-rc1"
+    SoftwareVersion.new(1, prerelease: "rc1", metadata: "build8").to_s.should eq "1-rc1+build8"
+  end
+
+  describe ".parse" do
+    it "parses metadata" do
+      SoftwareVersion.parse("1.0+1234")
+    end
+
+    it "valid values for 1.0" do
+      {"1.0", "1.0 ", " 1.0 ", "1.0\n", "\n1.0\n", "1.0"}.each do |good|
+        assert_version_equal "1.0", good
+      end
+    end
+
+    it "invalid values" do
+      invalid_versions = %w[
+        junk
+        1.0\n2.0
+        1..2
+        1.2\ 3.4
+      ]
+
+      # DON'T TOUCH THIS WITHOUT CHECKING CVE-2013-4287
+      invalid_versions << "2.3422222.222.222222222.22222.ads0as.dasd0.ddd2222.2.qd3e."
+
+      invalid_versions.each do |invalid|
+        expect_raises ArgumentError, "Malformed version string #{invalid.inspect}" do
+          SoftwareVersion.parse invalid
+        end
+        SoftwareVersion.parse?(invalid).should be_nil
+      end
+    end
+
+    it "empty version" do
+      ["", "   ", " "].each do |empty|
+        SoftwareVersion.parse(empty).to_s.should eq "0"
+      end
+    end
+  end
+
+  it "#<=>" do
+    # This spec has changed from Gems::Version where both where considered equal
+    v("1.0").should be < v("1.0.0")
+
+    v("1.0").should be > v("1.0.a")
+    v("1.8.2").should be > v("0.0.0")
+    v("1.8.2").should be > v("1.8.2.a")
+    v("1.8.2.b").should be > v("1.8.2.a")
+    v("1.8.2.a10").should be > v("1.8.2.a9")
+    v("1.0.0+build1").should be > v("1.0.0")
+    v("1.0.0+build2").should be > v("1.0.0+build1")
+
+    v("1.2.b1").should eq v("1.2.b.1")
+    v("").should eq v("0")
+    v("1.2.b1").should eq v("1.2.b1")
+    v("1.0a").should eq v("1.0.a")
+    v("1.0a").should eq v("1.0-a")
+    v("1.0.0+build1").should eq v("1.0.0+build1")
+
+    v("1.8.2.a").should be < v("1.8.2")
+    v("1.2").should be < v("1.3")
+    v("0.2.0").should be < v("0.2.0.1")
+    v("1.2.rc1").should be < v("1.2.rc2")
+  end
+
+  it "sort" do
+    list = ["0.1.0", "0.2.0", "5.333.1", "5.2.1", "0.2", "0.2.0.1", "5.8", "0.0.0.11"].map { |v| v(v) }
```

Comment:
Avoid creating a new array by modifying the existing array:
```suggestion
    list = ["0.1.0", "0.2.0", "5.333.1", "5.2.1", "0.2", "0.2.0.1", "5.8", "0.0.0.11"].map! { |v| v(v) }
```

Reply:
Well no, `v` call doesn't return a string..